<a href="https://colab.research.google.com/github/michalis0/Business-Intelligence-and-Analytics/blob/master/labs/12%20-%20Classification%202/exercises/exercises_classification_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<h1 align="center"> EXERCISES - CLASSIFICATION 2</h1>

<div>
<td>
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/2b/Logo_Universit%C3%A9_de_Lausanne.svg/2000px-Logo_Universit%C3%A9_de_Lausanne.svg.png" style="padding-right:10px;width:240px;float:left"/></td>
<h2 style="white-space: nowrap">Business Intelligence and Analytics</h2></td>
<hr style="clear:both">
<p style="font-size:0.85em; margin:2px; text-align:justify">

</div>

## KNN and Decision Trees

In this lab we will work on K Nearest Neighbor and Decision Trees for classification.

In [10]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import collections  as mc
%load_ext autoreload
%autoreload 2
import pandas as pd
import seaborn as sns
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
sns.set_style("white")

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [11]:
np.random.seed(72)

### Load data

We will start with the `heart disease` dataset from last week. Here's a description of the attributes in the data-set:

> 1. age
> 2. sex
> 3. chest pain type (4 values)
> 4. resting blood pressure
> 5. serum cholestoral in mg/dl
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved
> 9. exercise induced angina
> 10. target: 1: presence 0: absence of a heart disease

In [12]:
# data-set: heart.csv
df = pd.read_csv("https://media.githubusercontent.com/media/michalis0/Business-Intelligence-and-Analytics/master/data/heart_2025.csv", delimiter=';')
df.head()

age sex cp  trestbps  chol  fbs restecg  thalach exang  target
0   67   M  A       160   286   no       A      108   yes       0
1   67   M  A       120   229   no       A      129   yes       0
2   62   F  A       140   268   no       A      160    no       0
3   63   M  A       130   254   no       A      147    no       0
4   53   M  A       140   203  yes       A      155   yes       0

In [13]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import StandardScaler
from sklearn import datasets

### K-neirest neighbor classification

Let's start with only 2 features: age and maximum heart rate achieved (thalach). Define your features and target variable.

In [14]:
X = "YOUR CODE HERE"
y = "YOUR CODE HERE"

Split your data set into train and test subsets. Use the sklearn `train_test_split` with `test_size=0.2` and `random_state=72`.

In [15]:
from sklearn.model_selection import train_test_split
# YOUR CODE HERE

Define a KNN model with __3__ number of neighbors.


In [16]:
#Create model
model = "YOUR CODE HERE"

Fit your model now using the train set.

In [ ]:
model.fit("YOUR CODE HERE")

Compare your train and test accuracy for the KNN model.

In [ ]:
# train accuracy
"YOUR CODE HERE"

In [ ]:
# test accuracy
"YOUR CODE HERE"

Calculate the base rate

In [ ]:
#Compute the base rate
#Base rate = (most frequent class)/(total observations)
nbr_heart_disease = df.loc[df["YOUR CODE HERE"] == 1].shape[0]
print("#Heart disease = ", "YOUR CODE HERE")

nbr_no_heart_disease = df.loc[df["YOUR CODE HERE"] == 0].shape[0]
print("#No heart disease = ", "YOUR CODE HERE")

print("Baserate = ", "YOUR CODE HERE")

Show the confusion matrix for the KNN model.

In [ ]:
#Confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix("YOUR CODE HERE")

For which combination of K (number of neighbors), the value `p` (distance metric), and weight do we obtain the best accurary? Use the values indicated in the following code cell.

__Important: this exercise answers the Moodle quiz question 1.__

In [ ]:
n_neighbors = [2,3,4,5,6,7]
p_values = [1,2]
weights = ['uniform', 'distance']
for n in n_neighbors:
    for p in p_values:
        for w in weights:
            model = KNeighborsClassifier(n_neighbors=n, p=p, weights=w)
            model.fit("YOUR CODE HERE")
            print("accuracy for k = " + str(n) + ", p = " + str(p) + ", weight = " + w + " : " + str(model.score("YOUR CODE HERE")))

Now, using the KNN model with the best parameters you achieved above, let's try to predict the class and probability of correct classification for Age = 50, Thalach = 130 and Age = 70, Thalach = 160

In [ ]:
# define and train the model with the best parameters
model = KNeighborsClassifier("YOUR CODE HERE")
model.fit("YOUR CODE HERE")

In [ ]:
print(model.predict("YOUR CODE HERE"))
print(model.predict_proba("YOUR CODE HERE"))

Now let's try more numerical features and see if the accuracy improves.
We will use now "age", "thalach", "trestbps" and "chol".
Define your features and your target variable.

In [ ]:
X = "YOUR CODE HERE"
y = "YOUR CODE HERE"

Split your data set into train and test subsets. Again use `test_size=0.2, random_state=72`.

In [ ]:
# YOUR CODE HERE

Standardize your data using the sklearn standard scaler.

In [ ]:
standardize = StandardScaler()
"YOUR CODE HERE"

In [ ]:
X_train = "YOUR CODE HERE"
X_test = "YOUR CODE HERE"

Fit your model using the train data. Let's use the same KNN model that you defined above.



In [ ]:
model.fit("YOUR CODE HERE")

Compare your train and test accurary.

In [ ]:
# train accuracy
"YOUR CODE HERE"

In [ ]:
# test accuracy
"YOUR CODE HERE"

In [ ]:
#Confusion matrix for the test predictions
from sklearn.metrics import confusion_matrix
"YOUR CODE HERE"

**Question:** Compute the recall and precision of the model for the test set.

### Decision tree classification

We will now use the decision tree algorithm for classification. We will use the same dataset as the above exercise, i.e., we use the 4 features "age", "thalach", "trestbps", "chol".

Import the necessary libraries for decision tree classification, define the model and train your model. The test it. What is the accuracy of your model (using the test set)?

In [ ]:
from sklearn import tree
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(max_depth = 1)
model.fit("YOUR CODE HERE")
print(model.score("YOUR CODE HERE"))

How can you improve your model?

As you may know increasing the depth of the decision tree can improve its accuracy, but it can also lead to overfitting! Using the range of values for depth provided in the cell below, find the depth value that leads to the best test accuracy.

__Important: This exercise answers the Moodle quiz question 2.__

In [ ]:
best_acc = 0
for depth in range(1, 8):
  model = DecisionTreeClassifier("YOUR CODE HERE")
  model.fit("YOUR CODE HERE")
  accuracy = model.score("YOUR CODE HERE")
  if accuracy > best_acc:
    best_acc = accuracy
    best_depth = depth
print("best depth: ", "YOUR CODE HERE")